From 66c951c79771d0d86bb7917bc3428d2740162e7e Mon Sep 17 00:00:00 2001
From: Faisal Hassan <quic_faisalh@quicinc.com>
Date: Mon, 17 Mar 2025 16:57:05 +0530
Subject: [PATCH 09/17] usb: dwc3: qcom: Add support for sa8797p for qcom usb

QCOM SoC sa8797p allow multiple resources to be managed
by firmware. On this platform, tasks related to clocks, regulators,
resets, TLMM and interconnects can be delegated to the firmware,
while the remaining responsibilities are handled by Linux.

The power domain, managed through the PM runtime framework,
is utilized for activation/deactivation of the resources.

Power domain "usb_core" controls the GDSC, resets,
and TLMM(gpio) whereas power domain "usb_transfer"
controls clocks and interconnects.

In this variant, the two disjoint sets of resources are modeled against
the power domain suppliers which can be turned on/off to achieve modeled
device states. Two modeled low power states d1 and d3 are supported.

transition: d0 <--> d3  :: both core and transfer suppliers
transition: d0 <--> d1  :: only transfer supplier

Change-Id: If0b4602bbedffdee07550e74a24edf35ca616d63
Signed-off-by: Sriram Dash <quic_sriramd@quicinc.com>
Signed-off-by: Hiren Gohel <quic_vhiren@quicinc.com>
Signed-off-by: Faisal Hassan <quic_faisalh@quicinc.com>
Signed-off-by: Shazad Hussain <quic_shazhuss@quicinc.com>
---
 drivers/usb/dwc3/dwc3-qcom.c | 22 +++++++++++-----------
 1 file changed, 11 insertions(+), 11 deletions(-)

diff --git a/drivers/usb/dwc3/dwc3-qcom.c b/drivers/usb/dwc3/dwc3-qcom.c
index 965b333d3743..d8c8c9e26dcf 100644
--- a/drivers/usb/dwc3/dwc3-qcom.c
+++ b/drivers/usb/dwc3/dwc3-qcom.c
@@ -107,10 +107,6 @@ struct dwc3_qcom {
 
 #define to_dwc3_qcom(d) container_of((d), struct dwc3_qcom, dwc)
 
-static const struct dwc3_qcom_priv_data sa8255p_dwc3_qcom_priv_data = {
-	.fw_managed	= true,
-};
-
 static void dwc3_qcom_domain_detach(struct dwc3_qcom *qcom)
 {
 	dev_pm_domain_detach_list(qcom->pd_list);
@@ -120,7 +116,7 @@ static int dwc3_qcom_domain_attach(struct dwc3_qcom *qcom)
 {
 	struct dev_pm_domain_attach_data pd_data = {
 		.pd_flags	= PD_FLAG_NO_DEV_LINK,
-		.pd_names	= (const char*[]) { "usb_transfer", "usb_core" },
+		.pd_names	= (const char*[]) { "transfer", "core" },
 		.num_pd_names	= 2,
 	};
 	struct device *dev = qcom->dev;
@@ -143,11 +139,11 @@ static int dwc3_qcom_d3_to_d0(struct dwc3_qcom *qcom)
 	if (!qcom->pd_list)
 		return 0;
 
-	ret = pm_runtime_resume_and_get(qcom->pd_list->pd_devs[0]);
+	ret = pm_runtime_resume_and_get(qcom->pd_list->pd_devs[1]);
 	if (ret)
 		return ret;
 
-	return pm_runtime_resume_and_get(qcom->pd_list->pd_devs[1]);
+	return pm_runtime_resume_and_get(qcom->pd_list->pd_devs[0]);
 }
 
 /* d0_to_d3 transition by turning off all the suppliers */
@@ -741,11 +737,11 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	struct dwc3_probe_data	probe_data = {};
 	struct device		*dev = &pdev->dev;
 	struct dwc3_qcom	*qcom;
-	const struct dwc3_qcom_priv_data *priv_data;
 	struct resource		res;
 	struct resource		*r;
 	const struct dwc3_qcom_data *data;
 	int			ret;
+	bool			fw_managed;
 	bool			wakeup_source;
 
 	qcom = devm_kzalloc(&pdev->dev, sizeof(*qcom), GFP_KERNEL);
@@ -755,9 +751,9 @@ static int dwc3_qcom_probe(struct platform_device *pdev)
 	qcom->dev = &pdev->dev;
 	data = device_get_match_data(dev);
 
-	priv_data = of_device_get_match_data(dev);
+	fw_managed = device_get_match_data(dev);
 
-	if (priv_data && priv_data->fw_managed) {
+	if (fw_managed) {
 		ret = dwc3_qcom_domain_attach(qcom);
 		if (ret) {
 			dev_err(dev, "Failed to attach domains. Bail out\n");
@@ -1000,9 +996,13 @@ static const struct dev_pm_ops dwc3_qcom_dev_pm_ops = {
 
 static const struct of_device_id dwc3_qcom_of_match[] = {
 	{ .compatible = "qcom,nord-dwc3", .data = &dwc3_qcom_nord_data },
+	{
+		.compatible	= "qcom,sa8797p-dwc3",
+		.data		= (void *)true,
+	},
 	{
 		.compatible	= "qcom,sa8255p-dwc3",
-		.data		= &sa8255p_dwc3_qcom_priv_data,
+		.data		= (void *)true,
 	},
 	{ .compatible = "qcom,snps-dwc3" },
 	{ }
-- 
2.34.1

